File: frontend/src/api/alerts.ts
/**
 * API client for alerts system.
 */

import { DEFAULT_API_TIMEOUT_MS } from '../lib/apiConfig'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type AlertType =
  | 'price_above'
  | 'price_below'
  | 'rsi_overbought'
  | 'rsi_oversold'
  | 'macd_cross_up'
  | 'macd_cross_down'

export type ComparisonOperator = 'above' | 'below' | 'crosses_above' | 'crosses_below'

export type AlertCondition = {
  type: AlertType
  operator: ComparisonOperator
  value: number
  indicator_params?: Record<string, number>
}

export type Alert = {
  id: number
  symbol: string
  exchange: string
  timeframe: string
  condition_type: AlertType
  operator: ComparisonOperator
  threshold_value: number
  indicator_params: Record<string, number> | null
  enabled: boolean
  created_at: string
  triggered_at: string | null
  trigger_count: number
}

export type AlertHistory = {
  id: number
  alert_id: number
  triggered_at: string
  trigger_value: number
  price: number
  message: string
}

export type CreateAlertRequest = {
  symbol: string
  exchange: string
  timeframe: string
  condition: AlertCondition
  enabled?: boolean
}

export type UpdateAlertRequest = {
  enabled?: boolean
  threshold_value?: number
  indicator_params?: Record<string, number>
}

// ---------------------------------------------------------------------------
// API Functions
// ---------------------------------------------------------------------------

/**
 * Create a new alert.
 */
export async function createAlert(request: CreateAlertRequest): Promise<Alert> {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), DEFAULT_API_TIMEOUT_MS)

  try {
    const response = await fetch('/alerts/', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
      signal: controller.signal,
    })

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: 'Failed to create alert' }))
      throw new Error(error.detail || error.error || 'Failed to create alert')
    }

    return await response.json()
  } finally {
    clearTimeout(timeoutId)
  }
}

/**
 * List all alerts with optional filtering.
 */
export async function listAlerts(params?: {
  symbol?: string
  exchange?: string
  enabled_only?: boolean
}): Promise<{ alerts: Alert[]; count: number }> {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), DEFAULT_API_TIMEOUT_MS)

  try {
    const queryParams = new URLSearchParams()
    if (params?.symbol) queryParams.set('symbol', params.symbol)
    if (params?.exchange) queryParams.set('exchange', params.exchange)
    if (params?.enabled_only) queryParams.set('enabled_only', 'true')

    const url = `/alerts/${queryParams.toString() ? '?' + queryParams.toString() : ''}`
    const response = await fetch(url, { signal: controller.signal })

    if (!response.ok) {
      throw new Error('Failed to fetch alerts')
    }

    return await response.json()
  } finally {
    clearTimeout(timeoutId)
  }
}

/**
 * Get a specific alert by ID.
 */
export async function getAlert(alertId: number): Promise<Alert> {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), DEFAULT_API_TIMEOUT_MS)

  try {
    const response = await fetch(`/alerts/${alertId}`, { signal: controller.signal })

    if (!response.ok) {
      throw new Error('Alert not found')
    }

    return await response.json()
  } finally {
    clearTimeout(timeoutId)
  }
}

/**
 * Update an alert.
 */
export async function updateAlert(alertId: number, request: UpdateAlertRequest): Promise<Alert> {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), DEFAULT_API_TIMEOUT_MS)

  try {
    const response = await fetch(`/alerts/${alertId}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
      signal: controller.signal,
    })

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: 'Failed to update alert' }))
      throw new Error(error.detail || error.error || 'Failed to update alert')
    }

    return await response.json()
  } finally {
    clearTimeout(timeoutId)
  }
}

/**
 * Delete an alert.
 */
export async function deleteAlert(alertId: number): Promise<void> {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), DEFAULT_API_TIMEOUT_MS)

  try {
    const response = await fetch(`/alerts/${alertId}`, {
      method: 'DELETE',
      signal: controller.signal,
    })

    if (!response.ok) {
      throw new Error('Failed to delete alert')
    }
  } finally {
    clearTimeout(timeoutId)
  }
}

/**
 * Get alert history for a specific alert.
 */
export async function getAlertHistory(
  alertId: number,
  limit = 100
): Promise<{ history: AlertHistory[]; count: number }> {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), DEFAULT_API_TIMEOUT_MS)

  try {
    const response = await fetch(`/alerts/${alertId}/history?limit=${limit}`, {
      signal: controller.signal,
    })

    if (!response.ok) {
      throw new Error('Failed to fetch alert history')
    }

    return await response.json()
  } finally {
    clearTimeout(timeoutId)
  }
}

/**
 * Get all alert history.
 */
export async function getAllAlertHistory(
  limit = 100
): Promise<{ history: AlertHistory[]; count: number }> {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), DEFAULT_API_TIMEOUT_MS)

  try {
    const response = await fetch(`/alerts/history/all?limit=${limit}`, {
      signal: controller.signal,
    })

    if (!response.ok) {
      throw new Error('Failed to fetch alert history')
    }

    return await response.json()
  } finally {
    clearTimeout(timeoutId)
  }
}

File: api/routes/alerts.py
"""API routes for alert management."""

from __future__ import annotations

import logging
import os
from typing import Literal, Optional

import asyncpg
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from core.alerts.models import Alert, AlertHistory
from db.crud import alerts as alerts_crud

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/alerts", tags=["alerts"])


# ---------------------------------------------------------------------------
# Request/Response Models
# ---------------------------------------------------------------------------


class AlertConditionRequest(BaseModel):
    """Request model for alert condition."""

    type: Literal[
        "price_above",
        "price_below",
        "rsi_overbought",
        "rsi_oversold",
        "macd_cross_up",
        "macd_cross_down",
    ]
    operator: Literal["above", "below", "crosses_above", "crosses_below"]
    value: float
    indicator_params: Optional[dict] = None


class CreateAlertRequest(BaseModel):
    """Request to create a new alert."""

    symbol: str
    exchange: str
    timeframe: str
    condition: AlertConditionRequest
    enabled: bool = True


class UpdateAlertRequest(BaseModel):
    """Request to update an alert."""

    enabled: Optional[bool] = None
    threshold_value: Optional[float] = None
    indicator_params: Optional[dict] = None


class AlertResponse(BaseModel):
    """Response model for alert."""

    id: int
    symbol: str
    exchange: str
    timeframe: str
    condition_type: str
    operator: str
    threshold_value: float
    indicator_params: Optional[dict]
    enabled: bool
    created_at: str
    triggered_at: Optional[str]
    trigger_count: int


class AlertHistoryResponse(BaseModel):
    """Response model for alert history."""

    id: int
    alert_id: int
    triggered_at: str
    trigger_value: float
    price: float
    message: str


class ListAlertsResponse(BaseModel):
    """Response with list of alerts."""

    alerts: list[AlertResponse]
    count: int


class AlertHistoryListResponse(BaseModel):
    """Response with alert history."""

    history: list[AlertHistoryResponse]
    count: int


# ---------------------------------------------------------------------------
# Helper Functions
# ---------------------------------------------------------------------------


def _get_db_url() -> str:
    """Get database URL from environment."""
    db_url = os.getenv("DATABASE_URL", "postgresql://crypto:crypto@localhost:5432/cryptotrader")
    # Normalize to plain postgresql:// for asyncpg
    if db_url.startswith("postgresql+asyncpg://"):
        db_url = db_url.replace("postgresql+asyncpg://", "postgresql://")
    return db_url


async def _get_db_connection() -> asyncpg.Connection:
    """Get database connection."""
    return await asyncpg.connect(_get_db_url())


def _alert_to_response(alert: Alert) -> AlertResponse:
    """Convert Alert to AlertResponse."""
    return AlertResponse(
        id=alert.id or 0,
        symbol=alert.symbol,
        exchange=alert.exchange,
        timeframe=alert.timeframe,
        condition_type=alert.condition.type,
        operator=alert.condition.operator,
        threshold_value=alert.condition.value,
        indicator_params=alert.condition.indicator_params,
        enabled=alert.enabled,
        created_at=alert.created_at.isoformat() if alert.created_at else "",
        triggered_at=alert.triggered_at.isoformat() if alert.triggered_at else None,
        trigger_count=alert.trigger_count,
    )


def _history_to_response(history: AlertHistory) -> AlertHistoryResponse:
    """Convert AlertHistory to AlertHistoryResponse."""
    return AlertHistoryResponse(
        id=history.id or 0,
        alert_id=history.alert_id,
        triggered_at=history.triggered_at.isoformat(),
        trigger_value=history.trigger_value,
        price=history.price,
        message=history.message,
    )


# ---------------------------------------------------------------------------
# API Endpoints
# ---------------------------------------------------------------------------


@router.post("/", response_model=AlertResponse, status_code=201)
async def create_alert(request: CreateAlertRequest):
    """Create a new alert.

    Args:
        request: Alert creation request

    Returns:
        Created alert
    """
    conn = await _get_db_connection()
    try:
        alert = await alerts_crud.create_alert(
            conn,
            symbol=request.symbol,
            exchange=request.exchange,
            timeframe=request.timeframe,
            condition_type=request.condition.type,
            operator=request.condition.operator,
            threshold_value=request.condition.value,
            indicator_params=request.condition.indicator_params,
            enabled=request.enabled,
        )
        logger.info(f"Created alert {alert.id} for {request.symbol} on {request.exchange}")
        return _alert_to_response(alert)
    finally:
        await conn.close()


@router.get("/", response_model=ListAlertsResponse)
async def list_alerts(
    symbol: Optional[str] = None,
    exchange: Optional[str] = None,
    enabled_only: bool = False,
):
    """List alerts with optional filtering.

    Args:
        symbol: Optional symbol filter
        exchange: Optional exchange filter
        enabled_only: Only return enabled alerts

    Returns:
        List of alerts
    """
    conn = await _get_db_connection()
    try:
        alerts = await alerts_crud.list_alerts(
            conn,
            symbol=symbol,
            exchange=exchange,
            enabled_only=enabled_only,
        )
        return ListAlertsResponse(
            alerts=[_alert_to_response(alert) for alert in alerts],
            count=len(alerts),
        )
    finally:
        await conn.close()


@router.get("/{alert_id}", response_model=AlertResponse)
async def get_alert(alert_id: int):
    """Get a specific alert by ID.

    Args:
        alert_id: Alert ID

    Returns:
        Alert details

    Raises:
        HTTPException: If alert not found
    """
    conn = await _get_db_connection()
    try:
        alert = await alerts_crud.get_alert(conn, alert_id)
        if not alert:
            raise HTTPException(status_code=404, detail=f"Alert {alert_id} not found")
        return _alert_to_response(alert)
    finally:
        await conn.close()


@router.patch("/{alert_id}", response_model=AlertResponse)
async def update_alert(alert_id: int, request: UpdateAlertRequest):
    """Update an alert.

    Args:
        alert_id: Alert ID
        request: Update request

    Returns:
        Updated alert

    Raises:
        HTTPException: If alert not found
    """
    conn = await _get_db_connection()
    try:
        alert = await alerts_crud.update_alert(
            conn,
            alert_id,
            enabled=request.enabled,
            threshold_value=request.threshold_value,
            indicator_params=request.indicator_params,
        )
        if not alert:
            raise HTTPException(status_code=404, detail=f"Alert {alert_id} not found")
        logger.info(f"Updated alert {alert_id}")
        return _alert_to_response(alert)
    finally:
        await conn.close()


@router.delete("/{alert_id}", status_code=204)
async def delete_alert(alert_id: int):
    """Delete an alert.

    Args:
        alert_id: Alert ID

    Raises:
        HTTPException: If alert not found
    """
    conn = await _get_db_connection()
    try:
        success = await alerts_crud.delete_alert(conn, alert_id)
        if not success:
            raise HTTPException(status_code=404, detail=f"Alert {alert_id} not found")
        logger.info(f"Deleted alert {alert_id}")
    finally:
        await conn.close()


@router.get("/{alert_id}/history", response_model=AlertHistoryListResponse)
async def get_alert_history(alert_id: int, limit: int = 100):
    """Get history for a specific alert.

    Args:
        alert_id: Alert ID
        limit: Maximum number of records to return

    Returns:
        Alert history
    """
    conn = await _get_db_connection()
    try:
        history = await alerts_crud.get_alert_history(conn, alert_id=alert_id, limit=limit)
        return AlertHistoryListResponse(
            history=[_history_to_response(h) for h in history],
            count=len(history),
        )
    finally:
        await conn.close()


@router.get("/history/all", response_model=AlertHistoryListResponse)
async def get_all_alert_history(limit: int = 100):
    """Get history for all alerts.

    Args:
        limit: Maximum number of records to return

    Returns:
        Alert history across all alerts
    """
    conn = await _get_db_connection()
    try:
        history = await alerts_crud.get_alert_history(conn, limit=limit)
        return AlertHistoryListResponse(
            history=[_history_to_response(h) for h in history],
            count=len(history),
        )
    finally:
        await conn.close()
